myhdl/_module.py

```diff
@@ -0,0 +1,185 @@
+#  This file is part of the myhdl library, a Python package for using
+#  Python as a Hardware Description Language.
+#
+#  Copyright (C) 2003-2016 Jan Decaluwe
+#
+#  The myhdl library is free software; you can redistribute it and/or
+#  modify it under the terms of the GNU Lesser General Public License as
+#  published by the Free Software Foundation; either version 2.1 of the
+#  License, or (at your option) any later version.
+#
+#  This library is distributed in the hope that it will be useful, but
+#  WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  Lesser General Public License for more details.
+
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library; if not, write to the Free Software
+#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+
+""" Module with the @module decorator function. """
+from __future__ import absolute_import
+
+import inspect
+
+import myhdl
+from myhdl import ModuleError
+from myhdl._instance import _Instantiator
+from myhdl._util import _flatten
+from myhdl._extractHierarchy import (_MemInfo, _makeMemInfo,
+    _UserVerilogCode, _UserVhdlCode)
+from myhdl._Signal import _Signal, _isListOfSigs
+
+
+class _error:
+    pass
+_error.ArgType = "A module should return module or instantiator objects"
+_error.InstanceError = "%s: submodule %s should be encapsulated in a module decorator"
+
+class _CallInfo(object):
+    def __init__(self, name, modctxt, symdict):
+        self.name = name
+        self.modctxt = modctxt
+        self.symdict = symdict
+
+def _getCallInfo():
+    """Get info on the caller of a ModuleInstance.
+
+    A ModuleInstance should be used in a module context.
+    This function gets the required info from the caller
+    It uses the frame stack:
+    0: this function
+    1: module instance constructor
+    2: the _Module class __call__()
+    3: the function that defines instances
+    4: the caller of the module function, e.g. a ModuleInstance.
+
+    There is a complication when the decorator is used on a method.
+    In this case, it is used as a descriptor, and there is an additional
+    stack level due to the __get__ method. The current hack is to check
+    whether we are still in this module at level 3, and increment
+    all the subsequent levels.
+    """
+
+    stack = inspect.stack()
+    # check whether the decorator is used as a descriptor
+    if (inspect.getmodule(stack[3][0]) is myhdl._module):
+        funcrec = stack[4]
+        callerrec = stack[5]
+    else:
+        funcrec = stack[3]
+        callerrec = stack[4]
```

Comment:
This code needs to be more generic, there are cases when the stack size will vary.  See the comment in #123 for more information.